File: ios/Classes/Scanner Core/Scan Filters/CardNumberFilter.swift
//
//  CardNumberFilter.swift
//  card_scanner
//
//  Created by Abhinav Kumar on 12/02/21.
//

import Foundation
import MLKitTextRecognition

class CardNumberFilter: ScanFilter {
    private var cardNumberRegex: NSRegularExpression = try! NSRegularExpression(pattern: CardScannerRegexps.cardNumberRegex, options: [.anchorsMatchLines])
    
    var visionText: Text
    var scannerOptions: CardScannerOptions
    
    init(visionText: Text, scannerOptions: CardScannerOptions) {
        self.visionText = visionText
        self.scannerOptions = scannerOptions
    }
    
    func filter() -> ScanFilterResult? {
        for (blockIndex, block) in visionText.blocks.enumerated() {
            for (_, line) in block.lines.enumerated() {
                let sanitizedBlockText = line.text.sanitized
                debugLog("Sanitized Card Number : \(sanitizedBlockText)", scannerOptions: scannerOptions)
                
                if let firstMatch = cardNumberRegex.firstMatch(
                    in: sanitizedBlockText,
                    range: NSRange(location: 0, length: sanitizedBlockText.count)
                ) {
                    let cardNumber = (sanitizedBlockText as NSString).substring(with: firstMatch.range).trimmingCharacters(in: .whitespacesAndNewlines) as String
                    
                    if cardNumber.isNotValidCardNumber {
                        continue
                    }
                    
                    debugLog("Card Number : \(cardNumber)", scannerOptions: scannerOptions)
                    
                    if scannerOptions.enableLuhnCheck && cardNumber.isNotValidCardNumber {
                        debugLog("Luhn check failed for card number (\(cardNumber))", scannerOptions: scannerOptions)
                        continue
                    }
                    
                    return CardNumberScanResult(
                        visionText: visionText,
                        textBlockIndex: blockIndex,
                        textBlock: block,
                        cardNumber: cardNumber
                    )
                }
            }
        }
        
        return nil
    }
    
    
}

extension String {
    var isValidCardNumber: Bool {
        return true
    }
    
    var isNotValidCardNumber: Bool {
        return !isValidCardNumber
    }
    
    var isLuhnValidCardNumber: Bool {
        let digitList: [Int] = self.reversed().enumerated().map { (index, element) in
            var num = Int("\(element)") ?? 0
            if (index % 2 == 1) {
                num = (num * 2)
                num = (num == 0) ? num : (num % 9 == 0) ? 9 : num % 9
            }
            return num
        }
        
        return (digitList.reduce(0, +)) % 10 == 0
    }
    
    var isNotLuhnValidCardNumber: Bool {
        return !isLuhnValidCardNumber
    }
}

File: android/src/main/java/com/nateshmbhat/card_scanner/scanner_core/scan_filters/CardNumberFilter.kt
package com.nateshmbhat.card_scanner.scanner_core.scan_filters

import com.google.mlkit.vision.text.Text
import com.nateshmbhat.card_scanner.logger.debugLog
import com.nateshmbhat.card_scanner.scanner_core.constants.CardScannerRegexps
import com.nateshmbhat.card_scanner.scanner_core.models.CardNumberScanResult
import com.nateshmbhat.card_scanner.scanner_core.models.CardScannerOptions
import com.nateshmbhat.card_scanner.scanner_core.models.ScanFilter

class CardNumberFilter(visionText: Text, scannerOptions: CardScannerOptions) : ScanFilter(visionText, scannerOptions) {
  private val cardNumberRegex: Regex = Regex(CardScannerRegexps.cardNumberRegex, RegexOption.MULTILINE)

  fun _isValidCardNumber(cardNumber: String): Boolean {
    return true;
//    if (scannerOptions.customBinNumbersToScan?.isEmpty ?? true) {
//    return true;
//  } else {
//    for (var binNumber in scannerOptions.customBinNumbersToScan) {
//      if (cardNumber.startsWith(binNumber.trim()?.replaceAll(' ', ''))) return true;
//    }
//    return false;
//  }
  }

  override fun filter(): CardNumberScanResult? {
    for ((index, block) in visionText.textBlocks.withIndex()) {
      if (cardNumberRegex.containsMatchIn(block.text)) {
        val cardNumber = cardNumberRegex.find(block.text)!!.value.trim().replace(Regex("\\s+"), "")
        if (!_isValidCardNumber(cardNumber)) continue;
        debugLog("card number = $cardNumber", scannerOptions);
        if (scannerOptions.enableLuhnCheck && !checkLuhnAlgorithm(cardNumber)) {
          debugLog("Luhn check failed !", scannerOptions);
          continue;
        }
        return CardNumberScanResult(
                textBlockIndex = index, textBlock = block, cardNumber = cardNumber, visionText = visionText);
      }
    }
    return null;
  }

  ///[cleanedCardNumber] is card number without any extra space and with only the digits of the card
  private fun checkLuhnAlgorithm(cleanedCardNumber: String): Boolean {
    val digitList = cleanedCardNumber.reversed().mapIndexed { index, digit ->
      var num = "$digit".toInt()
      if (index % 2 == 1) {
        num = (num * 2)
        num = if (num == 0) num else if (num % 9 == 0) 9 else num % 9
      }
      num
    }

    return (digitList.sum()) % 10 == 0
  }
}